### DISCUSSION TITLE
Sonoff POW Elite 20A (POWR320D) - TotalEnergy counter gets reduced at times

### DISCUSSION THREAD

**Discussion by jkwim** [arendst/Tasmota#21999] [Q&A]
I am running Tasmota 14.1.0 on this device and sending the periodic measurements to an InfluxDB.
My visualization tool is Grafana.

When I look at the progressive increment of the counter TotalEnergy I see some dips ie. the value goes down at times.

The affected points are shown in the graph with yellow circles.

![image](https://github.com/user-attachments/assets/8dcbcdf7-3837-4bb2-97ff-ea7f98c48047)

I was running 12.1.3.4 and did an upgrade to 14.1.0 on 15th August 2024 (I have shown this also in the time line).

The issue remains.

When I look at the actual data stored in the database (database gets the data via MQTT), I can see the value jumping down. The strange thing that when this happens there is a small interval where data is not available. It is very likely that MQTT message was no sent by the device.
```

2024-07-14T15:59:51.782648326Z 2742.602
2024-07-14T16:00:51.750425691Z 2742.655
2024-07-14T16:01:51.771101711Z 2742.707
2024-07-14T16:02:51.791911387Z 2742.76
5 min gap?
2024-07-14T16:07:59.822928722Z 2720.628
2024-07-14T16:08:59.82257577Z  2720.628
2024-07-14T16:09:59.891509732Z 2720.628
2024-07-14T16:11:00.003953063Z 2720.628

```
```

2024-08-07T10:08:51.910314721Z 2817.881
2024-08-07T10:09:51.865854893Z 2817.881
2024-08-07T10:10:51.918234663Z 2817.881
2024-08-07T10:11:52.337618943Z 2817.881
4 min gap
2024-08-07T10:15:30.16179348Z  2806.314
2024-08-07T10:16:31.629472773Z 2806.314
2024-08-07T10:17:30.15692404Z  2806.314
2024-08-07T10:18:30.199510598Z 2806.314
2024-08-07T10:19:30.185928637Z 2806.314
2024-08-07T10:20:30.163597348Z 2806.314

```
```

2024-08-17T02:38:48.780942056Z 2846.349
2024-08-17T02:39:48.76880063Z  2846.349
15 min gap
2024-08-17T02:54:58.660035818Z 2845.356
2024-08-17T02:55:58.822602208Z 2845.356

```

```
2024-08-20T06:40:53.167607757Z 2862.538
2024-08-20T06:41:53.148936754Z 2862.538
2024-08-20T06:42:53.179046602Z 2862.538
2024-08-20T06:43:53.149599673Z 2862.538
2024-08-20T06:44:53.201441796Z 2862.538
3 min gap
2024-08-20T06:47:27.863073396Z 2849.77
2024-08-20T06:48:27.916576527Z 2849.77
2024-08-20T06:49:27.886516236Z 2849.77

```

Initially I thought that this may be associated at a timestamp when the power is being turned off. However examination of other incidents shows that it is not so. When the value remains constant ie. the load is switched off, also the jump can be seen.

Need some help to troubleshoot further please!

**Comment by sfromis** [comment]
What you post with those gaps does look "very consistent" with loss of input power (this is not the same as "power is being turned off", for the connected load), or maybe crashing, causing Tasmota to only have "last saved" values available after the restart.

**Reply by jkwim** [reply]
After posting this something came to my mind. The syslog should have the formatted message which will be fed to the InfluxDB.

So I looked and luckily found the logs for Aug 20th.


```
20 06:43:55,TotalEnergy=2862.538,YesterdayEnergy=0,TodayEnergy=12.768,Power=0,ApparentPower=0,ReactivePower=0,PowerFactor=0,Voltage=0,Current=0
20 06:44:55,TotalEnergy=2862.538,YesterdayEnergy=0,TodayEnergy=12.768,Power=0,ApparentPower=0,ReactivePower=0,PowerFactor=0,Voltage=0,Current=0
20 06:44:55,TotalEnergy=2862.538,YesterdayEnergy=0,TodayEnergy=12.768,Power=0,ApparentPower=0,ReactivePower=0,PowerFactor=0,Voltage=0,Current=0
20 06:45:55,TotalEnergy=2862.538,YesterdayEnergy=0,TodayEnergy=12.768,Power=0,ApparentPower=0,ReactivePower=0,PowerFactor=0,Voltage=0,Current=0
20 06:45:55,TotalEnergy=2862.538,YesterdayEnergy=0,TodayEnergy=12.768,Power=0,ApparentPower=0,ReactivePower=0,PowerFactor=0,Voltage=0,Current=0
20 06:46:55,TotalEnergy=2862.538,YesterdayEnergy=0,TodayEnergy=12.768,Power=0,ApparentPower=0,ReactivePower=0,PowerFactor=0,Voltage=0,Current=0
20 06:46:55,TotalEnergy=2862.538,YesterdayEnergy=0,TodayEnergy=12.768,Power=0,ApparentPower=0,ReactivePower=0,PowerFactor=0,Voltage=0,Current=0
20 06:47:55,TotalEnergy=2862.538,YesterdayEnergy=0,TodayEnergy=12.768,Power=0,ApparentPower=0,ReactivePower=0,PowerFactor=0,Voltage=0,Current=0
20 06:47:55,TotalEnergy=2862.538,YesterdayEnergy=0,TodayEnergy=12.768,Power=0,ApparentPower=0,ReactivePower=0,PowerFactor=0,Voltage=0,Current=0

```

So it appears that Tasmota is behaving correctly and sending the MQTT message properly. The messages have been received during the gap also.

Looks like a database issue.

**Reply by jkwim** [reply]
UPDATE:
My previous conclusion is wrong.

I did not consider the time zone offset in looking at the syslog.

Here is the correct log (I have truncated the columns for clarity)

```
20 12:13:56,TotalEnergy=2862.538,
20 12:13:56,TotalEnergy=2862.538,
20 12:14:56,TotalEnergy=2862.538,
20 12:14:56,TotalEnergy=2862.538,
   -----> 3 min gap
20 12:17:36,TotalEnergy=2849.77,Y
20 12:17:36,TotalEnergy=2849.77,Y
20 12:18:36,TotalEnergy=2849.77,Y
20 12:18:36,TotalEnergy=2849.77,Y
20 12:19:36,TotalEnergy=2849.77,Y
```
 
So Tasmota did not send the data during a 3min period and then started sending data with a lower value.

Now that suggests that Tasmota is not saving the current counter value in the flash for a long time right? In case of a restart, it picks from the last saved value in the flash?

**Reply by jkwim** [reply]
Here is the timestamp for the last saved value then:

```
20 01:28:54,TotalEnergy=2849.77,Y
20 01:29:53,TotalEnergy=2849.77,Y
20 01:29:54,TotalEnergy=2849.77,Y     <<<<<<<<<<<<<<<<<<
20 01:29:54,TotalEnergy=2849.771,
20 01:30:03,TotalEnergy=2849.771,
20 01:30:03,TotalEnergy=2849.772,
```

After 10 hrs and 45 minutes, the last saved value has been retrieved

```
20 12:17:36,TotalEnergy=2849.77,Y  <<<<<<<<<<<<
20 12:17:36,TotalEnergy=2849.77,Y
20 12:18:36,TotalEnergy=2849.77,Y
20 12:18:36,TotalEnergy=2849.77,Y
20 12:19:36,TotalEnergy=2849.77,Y
```

Shouldn't the "save to flash" interval be shorter?

**Reply by sfromis** [reply]
No. There is no "save to flash" interval at all, except at midnight.

It is deliberate that Tasmota is not wearing flash down by writing "all the time".

What I do for energy monitoring is to base it on what's published, and then I can compensate for outages, by counting my total up when the difference is not negative.

Of course, it is up to you if you want to incur the cost of more frequent writes, which you can trigger when you want, by issuing the `SaveData` command at pertinent times.

**Reply by jkwim** [reply]
Actually I wanted to use the TotalEnergy counter to plot a daily consumption graph like this:

![image](https://github.com/user-attachments/assets/3b5fef2f-b4ab-450e-8feb-0c8d4f049352)

The negative values distort the graph as shown.

I did not want to use the YesterdayEnergy and TodayEnergy counters as they get split at midnight and my required measurement interval crosses midnight.

Anyway things as it is I will see how I can circumvent the issue.

BTW, it is not clear to me how the SaveData command works:

```
SaveData 	
0 = save parameter changes only manually, e.g. with Restart 1
1 = save parameter changes every second (default)
2..3600 = save parameter changes every x second

```

When I checked current settings in the device:

```
10:52:42.946 CMD: savedata
10:52:43.047 MQT: stat/tasmota/SAVEDATA = {"SaveData":"ON"}

```

Does it mean that the data is saved every  second (ON=1)?

If it is the case then comes the question why the energy counter values did not get saved.

**Reply by sfromis** [reply]
No, it means that it is saved **once** when you issue the command. That's why I explained:
> by issuing the SaveData command **at pertinent times**.

Automatic saving is only when having changes of settings or power state, which is not all the time.

The "seconds" part of `SaveData` is basically to avoid multiple saves for multiple settings/state changes within an interval. Again, cutting down on flash writes (to the detriment of chip lifetime) is the key principle.
